File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    struct node{
        int key, val;
        node* prev;
        node* next;
        node(int k, int v): key(k), val(v), prev(nullptr), next(nullptr) {}
    };
    node* head = new node(-1, -1);
    node* tail = new node(-1, -1);

    int cap;
    unordered_map<int, node*> mp;
    void deleteAtEnd(node* delNode){
       node* prevNode = delNode->prev;
       node* nextNode = delNode->next;

       prevNode->next = nextNode;
       nextNode->prev = prevNode;
    }

    void insertAtStart(node* newNode){
        node* currentAfterHead = head->next;
        newNode->prev = head;
        newNode->next = currentAfterHead;
        currentAfterHead->prev = newNode;
        head->next = newNode;

    }
    LRUCache(int capacity) {
        cap = capacity;
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if(mp.find(key) != mp.end()){
            node* Node = mp[key];
            deleteAtEnd(Node);
            insertAtStart(Node);
            return Node->val;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(mp.find(key) != mp.end()){
            node* Node = mp[key];
            Node->val = value;
            deleteAtEnd(Node);
            insertAtStart(Node);
        }
        else{
            if(mp.size() == cap){
                node* Node = tail->prev;
                deleteAtEnd(Node);
                mp.erase(Node->key);
                
            }
            node* Node = new node(key, value);
            mp[key] = Node;
            insertAtStart(Node);          
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:

    def deleteNode(self, delNode):
        prevNode = delNode.prev
        nextNode = delNode.next
        prevNode.next = nextNode
        nextNode.prev = prevNode

    def insertAtStart(self, addNode):
        addNode.next = self.head.next
        addNode.prev = self.head

        self.head.next.prev = addNode
        self.head.next = addNode

    def __init__(self, capacity: int):
        self.map = {}
        self.cap = capacity
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key in self.map:
            node = self.map[key]
            self.deleteNode(node)
            self.insertAtStart(node)

            return node.val
        return -1


    def put(self, key: int, value: int) -> None:
        if key in self.map:
            node = self.map[key]
            node.val = value
            self.deleteNode(node)
            self.insertAtStart(node)
        else:
            if len(self.map) == self.cap:
                node = self.tail.prev
                self.deleteNode(node)
                del self.map[node.key]
            node = Node(key, value)
            self.map[key] = node
            self.insertAtStart(node)




# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)